SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Windows.Shared/Crashes.cs

```diff
@@ -224,33 +230,48 @@ private Task ProcessPendingErrorsAsync()
                         {
                             AppCenterLog.Warn(LogTag, $"Failed to delete error log file {file.Name}.", ex);
                         }
+                        continue;
+                    }
+                    var report = BuildErrorReport(log);
+                    if (report == null)
+                    {
+                        AppCenterLog.Error(LogTag, $"Error parsing error log. Deleting invalid file: {file.Name}");
+                        RemoveAllStoredErrorLogFiles(log.Id);
                     }
                     else
                     {
-                        _unprocessedManagedErrorLogs.Add(log.Id, log);
+                        if (lastSessionErrorReport == null || lastSessionErrorReport.AppErrorTime < report.AppErrorTime)
+                        {
+                            lastSessionErrorReport = report;
+                        }
+                        if (ShouldProcessErrorReport?.Invoke(report) ?? true)
+                        {
+                            _unprocessedManagedErrorLogs.Add(log.Id, log);
+                        }
+                        else
+                        {
+                            AppCenterLog.Debug(LogTag, $"ShouldProcessErrorReport returned false, clean up and ignore log: {log.Id}");
+                            RemoveAllStoredErrorLogFiles(log.Id);
+                        }
                     }
                 }
-                ErrorReport lastSessionErrorReport = null;
-                if (lastSessionErrorLog != null)
-                {
-                    AppCenterLog.Debug(LogTag, "Setting last session error report to an actual report.");
-
-                    // TODO: Build error report from cache.
-                    lastSessionErrorReport = new ErrorReport(lastSessionErrorLog, null);
-                }
-                else
-                {
-                    AppCenterLog.Debug(LogTag, "Setting last session error report to null.");
-                }
                 _lastSessionErrorReportTaskSource.SetResult(lastSessionErrorReport);
                 await SendCrashReportsOrAwaitUserConfirmationAsync().ConfigureAwait(false);
             });
         }
 
+        private void RemoveAllStoredErrorLogFiles(Guid errorId)
+        {
+            // ReSharper disable once InconsistentlySynchronizedField this is a concurrent dictionary.
+            _errorReportCache.Remove(errorId);
+            ErrorLogHelper.RemoveStoredErrorLogFile(errorId);
+            ErrorLogHelper.RemoveStoredExceptionFile(errorId);
+        }
+
         private async Task SendCrashReportsOrAwaitUserConfirmationAsync()
         {
-            bool alwaysSend = ApplicationSettings.GetValue(PrefKeyAlwaysSend, false);
-            if (_unprocessedManagedErrorLogs.Count() > 0)
+            var alwaysSend = ApplicationSettings.GetValue(PrefKeyAlwaysSend, false);
```

Comment:
Why less specific? Seems this should always be a bool.

Reply:
Always use var is the motto, you get a warning in resharper otherwise.
We have that in the entire code base unless we missed some occurrences.

Reply:
There are pros and cons but we are using `var` in all places if possible.